### ISSUE TITLE
TPM2 driver issue (code 10) in Windows10

### ISSUE THREAD

**Issue by Jeansen** [stefanberger/swtpm#669]
**Describe the bug**

**To Reproduce**
1. Create a windows10 vm in libvirt (vmm)
2. Add TPM (CRB or TIS)
3. Use EFI Secure Boot image
4.

I've attached my livbirt XML at the end.

**Expected behavior**
Windows runs with TPM2 enabled.

**Actual behavior***
Windows shows that it could not load the driver (code 10) in hardware manager

Driver used for TPM2 is installed automatically by Windows (Version 10.0.19041.746, Date 21.06.2006)

**Host (please complete the following information):**
 - OS: Debian 
 - Version bookworm/sid
 - Guest: Windows 10 build 19044

**Versions of relevant components**
- swtpm-build-deps/now 0.7.1 all 
swtpm-libs/unstable,now 0.7.1-1 amd64
swtpm-tools/unstable,now 0.7.1-1 amd64 
swtpm/unstable,now 0.7.1-1 amd64 
- openssl: OpenSSL 1.1.1m  14 Dec 2021
- gnutls:gnutls-bin/unstable,now 3.7.3-4+b1 amd64
- ...:

**Log files**
```
Successfully created RSA 2048 EK with handle 0x81010001.
  Invoking /usr/bin/swtpm_localca --type ek --ek dcd6a25f8a60fe38d2e2427105d0e32aa73c850bb29aae60a3372bfd3d5e8ff160d6dd5da0b676160fdf876c9c624368f94bd65a0d439effe038813c78a548e470feb735450748f2c585e3255a658a8faa1ecd94ac743801a068f548499c4fbd445802510597e012fe3f633b6151f4ec5613dbfa26f765a0b51af312fcf89bb8aede35ac281ce32af7ed979837adaa4a21af628b5ff9df1de8dd7191e2b02599706134a7e21bad24abb666885c48618a92a7ea542476e94beeb277f39adfe7eb2a1c6b9e0aa358783c13255646f586259aef7fe9dde3743f217568f821765e7d664ff3c01fd0d4c4cdb8a73aac83d691f89709ce4b5a12a666fa569dd071df67 --dir /tmp/swtpm_setup.certs.NU49H1 --logfile /var/log/swtpm/libvirt/qemu/win10-2-swtpm.log --vmid win10-2:dd77ed18-1178-4c84-b664-7b18adcb1e70 --tpm-spec-family 2.0 --tpm-spec-level 0 --tpm-spec-revision 164 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:20191023 --tpm2 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options
Successfully created EK certificate locally.
  Invoking /usr/bin/swtpm_localca --type platform --ek dcd6a25f8a60fe38d2e2427105d0e32aa73c850bb29aae60a3372bfd3d5e8ff160d6dd5da0b676160fdf876c9c624368f94bd65a0d439effe038813c78a548e470feb735450748f2c585e3255a658a8faa1ecd94ac743801a068f548499c4fbd445802510597e012fe3f633b6151f4ec5613dbfa26f765a0b51af312fcf89bb8aede35ac281ce32af7ed979837adaa4a21af628b5ff9df1de8dd7191e2b02599706134a7e21bad24abb666885c48618a92a7ea542476e94beeb277f39adfe7eb2a1c6b9e0aa358783c13255646f586259aef7fe9dde3743f217568f821765e7d664ff3c01fd0d4c4cdb8a73aac83d691f89709ce4b5a12a666fa569dd071df67 --dir /tmp/swtpm_setup.certs.NU49H1 --logfile /var/log/swtpm/libvirt/qemu/win10-2-swtpm.log --vmid win10-2:dd77ed18-1178-4c84-b664-7b18adcb1e70 --tpm-spec-family 2.0 --tpm-spec-level 0 --tpm-spec-revision 164 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:20191023 --tpm2 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options
Successfully created platform certificate locally.
Successfully created NVRAM area 0x1c00002 for RSA 2048 EK certificate.
Successfully created NVRAM area 0x1c08000 for platform certificate.
Successfully created ECC EK with handle 0x81010016.
  Invoking /usr/bin/swtpm_localca --type ek --ek x=3bb2b747ed3ab537d7c8cbdb59b6cedebe7276cf47ebd04564729a3ac8f0a9075b3b2d38e9b542eaaff4fa12b8777dac,y=86d2b056042e3d0f021b1e6c49f0da48eb36006440252ed08850779a39ed8c2615f09350424bc2580dad145632e1447f,id=secp384r1 --dir /tmp/swtpm_setup.certs.NU49H1 --logfile /var/log/swtpm/libvirt/qemu/win10-2-swtpm.log --vmid win10-2:dd77ed18-1178-4c84-b664-7b18adcb1e70 --tpm-spec-family 2.0 --tpm-spec-level 0 --tpm-spec-revision 164 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:20191023 --tpm2 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options
Successfully created EK certificate locally.
Successfully created NVRAM area 0x1c00016 for ECC EK certificate.
Successfully activated PCR banks sha256 among sha1,sha256,sha384,sha512.
Successfully authored TPM state.
Ending vTPM manufacturing @ Sun 27 Feb 2022 11:44:54 AM CET
```

**Additional context**
Libvirt exectuion from logfile:
```
Successfully created RSA 2048 EK with handle 0x81010001.
  Invoking /usr/bin/swtpm_localca --type ek --ek dcd6a25f8a60fe38d2e2427105d0e32aa73c850bb29aae60a3372bfd3d5e8ff160d6dd5da0b676160fdf876c9c624368f94bd65a0d439effe038813c78a548e470feb735450748f2c585e3255a658a8faa1ecd94ac743801a068f548499c4fbd445802510597e012fe3f633b6151f4ec5613dbfa26f765a0b51af312fcf89bb8aede35ac281ce32af7ed979837adaa4a21af628b5ff9df1de8dd7191e2b02599706134a7e21bad24abb666885c48618a92a7ea542476e94beeb277f39adfe7eb2a1c6b9e0aa358783c13255646f586259aef7fe9dde3743f217568f821765e7d664ff3c01fd0d4c4cdb8a73aac83d691f89709ce4b5a12a666fa569dd071df67 --dir /tmp/swtpm_setup.certs.NU49H1 --logfile /var/log/swtpm/libvirt/qemu/win10-2-swtpm.log --vmid win10-2:dd77ed18-1178-4c84-b664-7b18adcb1e70 --tpm-spec-family 2.0 --tpm-spec-level 0 --tpm-spec-revision 164 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:20191023 --tpm2 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options
Successfully created EK certificate locally.
  Invoking /usr/bin/swtpm_localca --type platform --ek dcd6a25f8a60fe38d2e2427105d0e32aa73c850bb29aae60a3372bfd3d5e8ff160d6dd5da0b676160fdf876c9c624368f94bd65a0d439effe038813c78a548e470feb735450748f2c585e3255a658a8faa1ecd94ac743801a068f548499c4fbd445802510597e012fe3f633b6151f4ec5613dbfa26f765a0b51af312fcf89bb8aede35ac281ce32af7ed979837adaa4a21af628b5ff9df1de8dd7191e2b02599706134a7e21bad24abb666885c48618a92a7ea542476e94beeb277f39adfe7eb2a1c6b9e0aa358783c13255646f586259aef7fe9dde3743f217568f821765e7d664ff3c01fd0d4c4cdb8a73aac83d691f89709ce4b5a12a666fa569dd071df67 --dir /tmp/swtpm_setup.certs.NU49H1 --logfile /var/log/swtpm/libvirt/qemu/win10-2-swtpm.log --vmid win10-2:dd77ed18-1178-4c84-b664-7b18adcb1e70 --tpm-spec-family 2.0 --tpm-spec-level 0 --tpm-spec-revision 164 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:20191023 --tpm2 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options
Successfully created platform certificate locally.
Successfully created NVRAM area 0x1c00002 for RSA 2048 EK certificate.
Successfully created NVRAM area 0x1c08000 for platform certificate.
Successfully created ECC EK with handle 0x81010016.
  Invoking /usr/bin/swtpm_localca --type ek --ek x=3bb2b747ed3ab537d7c8cbdb59b6cedebe7276cf47ebd04564729a3ac8f0a9075b3b2d38e9b542eaaff4fa12b8777dac,y=86d2b056042e3d0f021b1e6c49f0da48eb36006440252ed08850779a39ed8c2615f09350424bc2580dad145632e1447f,id=secp384r1 --dir /tmp/swtpm_setup.certs.NU49H1 --logfile /var/log/swtpm/libvirt/qemu/win10-2-swtpm.log --vmid win10-2:dd77ed18-1178-4c84-b664-7b18adcb1e70 --tpm-spec-family 2.0 --tpm-spec-level 0 --tpm-spec-revision 164 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:20191023 --tpm2 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options
Successfully created EK certificate locally.
Successfully created NVRAM area 0x1c00016 for ECC EK certificate.
Successfully activated PCR banks sha256 among sha1,sha256,sha384,sha512.
Successfully authored TPM state.
Ending vTPM manufacturing @ Sun 27 Feb 2022 11:44:54 AM CET
```

libvirt xml config:
```
<domain type="kvm">
  <name>win10-2</name>
  <uuid>dd77ed18-1178-4c84-b664-7b18adcb1e70</uuid>
  <metadata>
    <libosinfo:libosinfo xmlns:libosinfo="http://libosinfo.org/xmlns/libvirt/domain/1.0">
      <libosinfo:os id="http://microsoft.com/win/10"/>
    </libosinfo:libosinfo>
  </metadata>
  <memory unit="KiB">8388608</memory>
  <currentMemory unit="KiB">8388608</currentMemory>
  <vcpu placement="static">4</vcpu>
  <os>
    <type arch="x86_64" machine="pc-q35-6.2">hvm</type>
    <loader readonly="yes" secure="yes" type="pflash">/usr/share/OVMF/OVMF_CODE_4M.secboot.fd</loader>
    <nvram>/var/lib/libvirt/qemu/nvram/win10-2_VARS.fd</nvram>
    <boot dev="hd"/>
  </os>
  <features>
    <acpi/>
    <apic/>
    <hyperv mode="custom">
      <relaxed state="on"/>
      <vapic state="on"/>
      <spinlocks state="on" retries="8191"/>
    </hyperv>
    <vmport state="off"/>
    <smm state="on"/>
  </features>
  <cpu mode="host-passthrough" check="none" migratable="on"/>
  <clock offset="localtime">
    <timer name="rtc" tickpolicy="catchup"/>
    <timer name="pit" tickpolicy="delay"/>
    <timer name="hpet" present="no"/>
    <timer name="hypervclock" present="yes"/>
  </clock>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <pm>
    <suspend-to-mem enabled="no"/>
    <suspend-to-disk enabled="no"/>
  </pm>
  <devices>
    <emulator>/usr/bin/qemu-system-x86_64</emulator>
    <disk type="file" device="disk">
      <driver name="qemu" type="qcow2"/>
      <source file="/home/marcel/libvirt/Windows_10.qcow2"/>
      <target dev="vda" bus="virtio"/>
      <address type="pci" domain="0x0000" bus="0x04" slot="0x00" function="0x0"/>
    </disk>
    <controller type="usb" index="0" model="qemu-xhci" ports="15">
      <address type="pci" domain="0x0000" bus="0x02" slot="0x00" function="0x0"/>
    </controller>
    <controller type="sata" index="0">
      <address type="pci" domain="0x0000" bus="0x00" slot="0x1f" function="0x2"/>
    </controller>
    <controller type="pci" index="0" model="pcie-root"/>
    <controller type="pci" index="1" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="1" port="0x10"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x0" multifunction="on"/>
    </controller>
    <controller type="pci" index="2" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="2" port="0x11"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x1"/>
    </controller>
    <controller type="pci" index="3" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="3" port="0x12"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x2"/>
    </controller>
    <controller type="pci" index="4" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="4" port="0x13"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x3"/>
    </controller>
    <controller type="pci" index="5" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="5" port="0x14"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x4"/>
    </controller>
    <controller type="pci" index="6" model="pcie-root-port">
      <model name="pcie-root-port"/>
      <target chassis="6" port="0x15"/>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x02" function="0x5"/>
    </controller>
    <controller type="virtio-serial" index="0">
      <address type="pci" domain="0x0000" bus="0x03" slot="0x00" function="0x0"/>
    </controller>
    <interface type="network">
      <mac address="52:54:00:0c:46:b3"/>
      <source network="default"/>
      <model type="e1000e"/>
      <address type="pci" domain="0x0000" bus="0x01" slot="0x00" function="0x0"/>
    </interface>
    <serial type="pty">
      <target type="isa-serial" port="0">
        <model name="isa-serial"/>
      </target>
    </serial>
    <console type="pty">
      <target type="serial" port="0"/>
    </console>
    <channel type="spicevmc">
      <target type="virtio" name="com.redhat.spice.0"/>
      <address type="virtio-serial" controller="0" bus="0" port="1"/>
    </channel>
    <input type="tablet" bus="usb">
      <address type="usb" bus="0" port="1"/>
    </input>
    <input type="mouse" bus="ps2"/>
    <input type="keyboard" bus="ps2"/>
    <tpm model="tpm-crb">
      <backend type="emulator" version="2.0"/>
    </tpm>
    <graphics type="spice">
      <listen type="none"/>
      <image compression="off"/>
      <gl enable="yes" rendernode="/dev/dri/by-path/pci-0000:00:02.0-render"/>
    </graphics>
    <sound model="ich9">
      <address type="pci" domain="0x0000" bus="0x00" slot="0x1b" function="0x0"/>
    </sound>
    <audio id="1" type="spice"/>
    <video>
      <model type="virtio" heads="1" primary="yes">
        <acceleration accel3d="yes"/>
      </model>
      <address type="pci" domain="0x0000" bus="0x00" slot="0x01" function="0x0"/>
    </video>
    <redirdev bus="usb" type="spicevmc">
      <address type="usb" bus="0" port="2"/>
    </redirdev>
    <redirdev bus="usb" type="spicevmc">
      <address type="usb" bus="0" port="3"/>
    </redirdev>
    <memballoon model="virtio">
      <address type="pci" domain="0x0000" bus="0x05" slot="0x00" function="0x0"/>
    </memballoon>
  </devices>
</domain>

```

**Comment by stefanberger** [comment]
> Windows shows that it could not load the driver (code 10) in hardware manager

What is the error message?

Also, what version of QEMU did you use?

**Comment by Jeansen** [comment]
@stefanberger Thanks for the quick reply.

QEMU:
QEMU emulator version 6.2.0 (Debian 1:6.2+dfsg-3)
Copyright (c) 2003-2021 Fabrice Bellard and the QEMU Project developers

Code 10 ist documented here:
"This device cannot start. Try upgrading the device drivers for this device. (Code 10)"
https://support.microsoft.com/en-us/topic/error-codes-in-device-manager-in-windows-524e9e89-4dee-8883-0afa-6bca0456324e

**Comment by stefanberger** [comment]
Can you please play around with the hardware configuration of the VM and remove / change a couple of devices to see whether it makes a difference. Like the 3D accelerator that I wouldn't be able to test with. Win10 I think has been tried by quite a few people by now...

**Comment by Jeansen** [comment]
OK, I did play around with about a dozen reboots. Worked my way from disabling 3D until I ended up with SPICE. Changed RAM, removed e.g. tablet input and even changed the NIC. I always then checked the hardware manager. Uninstalled the TPM driver and had Windows recheck for new hardware so it would find the TPM device. It always finds it, but then cries regarding the driver load.

**Comment by stefanberger** [comment]
I have now tried 4 win10 installations with QEMU 6.1 and QEMU 6.2 with SeaBIOS and UEFI-secure-boot each. All of them seem to be able to use the TPM 2.0 device just fine.

> Driver used for TPM2 is installed automatically by Windows (Version 10.0.19041.746, Date 21.06.2006)
I have it working with the same driver.

Here's my libvirt XML:

```
<domain type='kvm'>
  <name>win10-qemu-6.2-uefi</name>
  <uuid>461e8781-310d-4f4e-b13c-b9de7d7b54ff</uuid>
  <metadata>
    <libosinfo:libosinfo xmlns:libosinfo="http://libosinfo.org/xmlns/libvirt/domain/1.0">
      <libosinfo:os id="http://microsoft.com/win/10"/>
    </libosinfo:libosinfo>
  </metadata>
  <memory unit='KiB'>4194304</memory>
  <currentMemory unit='KiB'>4194304</currentMemory>
  <vcpu placement='static'>8</vcpu>
  <os>
    <type arch='x86_64' machine='pc-q35-6.2'>hvm</type>
    <loader readonly='yes' type='pflash'>/usr/share/qemu/edk2-x86_64-secure-code.fd</loader>
    <nvram>/var/lib/libvirt/qemu/nvram/win10-qemu-6.2-uefi_VARS.fd</nvram>
    <bootmenu enable='yes'/>
  </os>
  <features>
    <acpi/>
    <apic/>
    <hyperv>
      <relaxed state='on'/>
      <vapic state='on'/>
      <spinlocks state='on' retries='8191'/>
    </hyperv>
    <vmport state='off'/>
  </features>
  <cpu mode='host-model' check='partial'/>
  <clock offset='localtime'>
    <timer name='rtc' tickpolicy='catchup'/>
    <timer name='pit' tickpolicy='delay'/>
    <timer name='hpet' present='no'/>
    <timer name='hypervclock' present='yes'/>
  </clock>
  <on_poweroff>destroy</on_poweroff>
  <on_reboot>restart</on_reboot>
  <on_crash>destroy</on_crash>
  <pm>
    <suspend-to-mem enabled='no'/>
    <suspend-to-disk enabled='no'/>
  </pm>
  <devices>
    <emulator>/usr/local/bin/qemu-system-x86_64</emulator>
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/var/lib/libvirt/images/win10-qemu-6.2-uefi.qcow2'/>
      <target dev='sda' bus='sata'/>
      <boot order='1'/>
      <address type='drive' controller='0' bus='0' target='0' unit='0'/>
    </disk>
    <disk type='file' device='cdrom'>
      <driver name='qemu' type='raw'/>
      <target dev='sdb' bus='sata'/>
      <readonly/>
      <boot order='2'/>
      <address type='drive' controller='0' bus='0' target='0' unit='1'/>
    </disk>
    <controller type='usb' index='0' model='qemu-xhci' ports='15'>
      <address type='pci' domain='0x0000' bus='0x02' slot='0x00' function='0x0'/>
    </controller>
    <controller type='sata' index='0'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x1f' function='0x2'/>
    </controller>
    <controller type='pci' index='0' model='pcie-root'/>
    <controller type='pci' index='1' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='1' port='0x10'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0' multifunction='on'/>
    </controller>
    <controller type='pci' index='2' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='2' port='0x11'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x1'/>
    </controller>
    <controller type='pci' index='3' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='3' port='0x12'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x2'/>
    </controller>
    <controller type='pci' index='4' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='4' port='0x13'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x3'/>
    </controller>
    <controller type='pci' index='5' model='pcie-root-port'>
      <model name='pcie-root-port'/>
      <target chassis='5' port='0x14'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x4'/>
    </controller>
    <controller type='virtio-serial' index='0'>
      <address type='pci' domain='0x0000' bus='0x03' slot='0x00' function='0x0'/>
    </controller>
    <interface type='network'>
      <mac address='52:54:00:57:90:a0'/>
      <source network='test'/>
      <model type='e1000e'/>
      <address type='pci' domain='0x0000' bus='0x01' slot='0x00' function='0x0'/>
    </interface>
    <serial type='pty'>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
    </serial>
    <console type='pty'>
      <target type='serial' port='0'/>
    </console>
    <channel type='spicevmc'>
      <target type='virtio' name='com.redhat.spice.0'/>
      <address type='virtio-serial' controller='0' bus='0' port='1'/>
    </channel>
    <input type='tablet' bus='usb'>
      <address type='usb' bus='0' port='1'/>
    </input>
    <input type='mouse' bus='ps2'/>
    <input type='keyboard' bus='ps2'/>
    <tpm model='tpm-crb'>
      <backend type='emulator' version='2.0'/>
    </tpm>
    <graphics type='spice' autoport='yes'>
      <listen type='address'/>
    </graphics>
    <sound model='ich9'>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x1b' function='0x0'/>
    </sound>
    <audio id='1' type='spice'/>
    <video>
      <model type='qxl' ram='65536' vram='65536' vgamem='16384' heads='1' primary='yes'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x0'/>
    </video>
    <redirdev bus='usb' type='spicevmc'>
      <address type='usb' bus='0' port='2'/>
    </redirdev>
    <redirdev bus='usb' type='spicevmc'>
      <address type='usb' bus='0' port='3'/>
    </redirdev>
    <memballoon model='virtio'>
      <address type='pci' domain='0x0000' bus='0x04' slot='0x00' function='0x0'/>
    </memballoon>
  </devices>
</domain>
```

If it's a driver issue I would think it's an issue in QEMU (interface or ACPI), but that doesn't seem to be the case here.

**Comment by stefanberger** [comment]
Are there any errors reported in the swtpm log (`/var/log/swtpm/libvirt/...`)? Does it help to put AppArmor in complain mode or disable it? Can you enter the UEFI TPM 2 menu (Device Manager->TCG2 Configuration)?

**Comment by Jeansen** [comment]
No, swtpm logs are filled with successfully entries. But it must be somewhere on my side then ... I've taken the day to roll back some snapshots even until half a year ago. I know it worked somewhere in-between. I'll try some more tomorrow. But since this is obviously an issue on my side, I am closing this one. I hope, testing on your side did not take too much time. I'll post my findings - if any - here for reference.

**Comment by Jeansen** [comment]
OK, I give up. I reset my Windows to a clean install, removed all debian package traces and worked my way down from manually compiling 0.7.1 down to 6.1. It's always the same. Disabling AppArmor did not help. I even reinstalled the ovmf package which provides the efi imaes. Sadly, I cannot see any traces in the log and my only guess is that it mus have to do with my libvirt/kvm/qemu versions ....

**Comment by stefanberger** [comment]
Do you have a Linux VM running?

**Comment by stefanberger** [comment]
Are you using libtpms v0.9.2? I have a Ubuntu repo here that may work with Debian as well, but I am not sure since I don't typically use Debian: https://launchpad.net/~stefanberger/+archive/ubuntu/swtpm-focal

**Comment by Jeansen** [comment]
Yes, It's 0.9.2. I built with the debian packaga as well as with your package that I build manually from your github repo - as referenced in the wiki page.

**Comment by Jeansen** [comment]
When I find some more time, I think I will run it on a Debian stable installation and see what happens there. But this will need some time to prep. Or it will be gone when the next version of Qemu arrives. Maybe with 7.0 things change .... After all, I am not on the unstable ring with my current setup....

**Comment by linsui** [comment]
I have the same problem. TPM 1.2 works well.

**Comment by ncopa** [comment]
I wonder if this can be re-opened or if I should create a new issue? I have a similar issue here with Alpine Linux.

```
ncopa-desktop:~$ qemu-system-x86_64 --version
QEMU emulator version 7.1.0
Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers
ncopa-desktop:~$ swtpm --version
TPM emulator version 0.7.3, Copyright (c) 2014-2021 IBM Corp.
ncopa-desktop:~$ virt-manager --version
4.1.0
ncopa-desktop:~$ virsh --version
8.7.0
```

In windows 10 device manager I have this error:
```
This device cannot start. (Code 10)

A protocol error was detected between the driver and the device.
```

swtpm is built with openssl3:
```sh
build() {
        ./configure \
                --build=$CBUILD \
                --host=$CHOST \
                --prefix=/usr \
                --libdir=/usr/lib \
                --with-openssl \
                --with-tss-user=root \
                --with-tss-group=root
        make
}
```

I am able to test different builds and patches, but I have no clue how to troubleshoot it.

**Comment by ncopa** [comment]
I have tested with both TIS and CRB in virt-manager/libvirt/qemu.

**Comment by Jeansen** [comment]
What version of this driver do you use? For me, after some updates the issues were gone (either by updates of this driver or libvirt).

**Comment by Jeansen** [comment]
I have 0.7.1 from the Debian repos in use. Works flawlessly.

**Comment by ncopa** [comment]
```
$ swtpm --version
TPM emulator version 0.7.3, Copyright (c) 2014-2021 IBM Corp.
```

I have tested with libtpms-0.9.5 and libtpms-0.9.3.

**Comment by ncopa** [comment]
Now have I also tested with (unpatched)
`TPM emulator version 0.7.1, Copyright (c) 2014-2021 IBM Corp.`

Still not working.

**Comment by ncopa** [comment]
> I have 0.7.1 from the Debian repos in use. Works flawlessly.

Seems like debian's 0.7.1 does not have any patches (other than typos in docs and help text):
https://salsa.debian.org/debian/swtpm/-/tree/master/debian/patches

So, I would think this means that problem is not in swtpm but somewhere else. Maybe in libtpms?

**Comment by stefanberger** [comment]
I have not seen this type of problem... Does your UEFI show that the TPM 2 is recognized -- it should probably have a menu entry. Does Linux work in the VM?

**Comment by ncopa** [comment]
> I have not seen this type of problem... Does your UEFI show that the TPM 2 is recognized -- it should probably have a menu entry.

I think I found it. It is the UEFI that is broken.

Some references:
- https://groups.google.com/g/linux.debian.bugs.dist/c/KIARY7bL5RI
- edk2 upstream change that triggered the regression https://github.com/tianocore/edk2/commit/4de8d61bcec02a13ceed84f92b0cf3ea58adf9c5
- debian fix: https://salsa.debian.org/qemu-team/edk2/-/commit/adf1a2ec7fd7f6488cfec59fe5a8e3054ba16b75

> Does Linux work in the VM?

Linux works in the VM but I haven't checked if tpm 2.0 works in linux.

I rebuilt Alpine's edk2/ovmf package with `-DSECURE_BOOT_ENABLE=TRUE` and `-DTPM2_ENABLE=TRUE` and when configuring the VM to use this UEFI firmware it works.

I believe this means that the  `/usr/share/qemu/edk2-x86_64-secure-code.fd` firmware shipped with qemu 7.1.0 is broken, probably due to the [edk2 upstream change](https://github.com/tianocore/edk2/commit/4de8d61bcec02a13ceed84f92b0cf3ea58adf9c5).
I'll investigate this and follow up with qemu devs.

Thank you!

**Comment by stefanberger** [comment]
> I believe this means that the `/usr/share/qemu/edk2-x86_64-secure-code.fd` firmware shipped with qemu 7.1.0 is broken, probably due to the [edk2 upstream change](https://github.com/tianocore/edk2/commit/4de8d61bcec02a13ceed84f92b0cf3ea58adf9c5).

That's right. There was this switch in EDK2 from some TPM-related CONFIG options that were not used when compiling EDK2 for QEMU. I sent a patch to the qemu devel mailing list fixing it:

https://lists.nongnu.org/archive/html/qemu-devel/2022-06/msg01441.html
